Add support for elaborated type specifiers

iMichka opened the pull request:
Elaborated type specifiers may be used to refer to a previously-declared
class name (class, struct, or union) or to a previously-declared enum name.

These type specifiers are important for C code, but are also sometimes
relavant in the context of C++ code.

This patch adds the ElaboratedType xml node.

The castxml output format version number was updated to 1.1.0.

A test was added to demonstrate the purpose of this change (test/input/Elaborated.cxx).

iMichka commented:
@bradking I rebased the branch and fixed the conflicts.

bradking commented on src/Output.cxx:
```diff
@@ -732,10 +741,6 @@ ASTVisitor::DumpId ASTVisitor::AddTypeDumpNode(DumpType dt, bool complete,
-    case clang::Type::Elaborated:
-      return this->AddTypeDumpNode(
-        DumpType(t->getAs<clang::ElaboratedType>()->getNamedType(), c),
-        complete, dq);
```
We should be able to keep the logic in this `case` and just `break` when it is not gccxml mode.

bradking commented on test/input/Elaborated.cxx:
```diff
@@ -0,0 +1,49 @@
+    void func3(Foo3 *a5, union Foo3 *a6);
+    void func4(Foo4 *a7, class Foo4 *a8);
+
+}
```
Please add a newline at the end of the last line.

iMichka commented:
Done:
- newline added
- if condition for elaborated type was moved back to the case block
- the difference between c++98 and c++11 was due to the throw attributes. I merged both files and am using `( throw="")?` now.
- I added a test file with the output for the old file format

bradking commented:
Thanks.  I updated your branch with a couple tweaks.  Please reset your local checkout.

bradking commented on test/expect/castxml1.any.qualified-type-name.xml.txt:
```diff
@@ -1,9 +1,10 @@
+  <Typedef id="_4" name="type" type="_5" context="_6" location="f1:2" file="f1" line="2"/>
+  <FundamentalType id="_5" name="int" size="[0-9]+" align="[0-9]+"/>
+  <Namespace id="_6" name="ns" context="_3"/>
   <File id="f1" name=".*/test/input/qualified-type-name.cxx"/>
```
The appearance of `ElaboratedType` here is not due to `struct foo` but instead to qualification like `ns::type`.  This information may or may not be useful to clients, but currently `ElaboratedType` elements do not indicate what kind of elaboration was used.

iMichka replied:
I am not sure what you mean with: `what kind of elaboration was used`.

`<ElaboratedType id="_2" type="_4"/>` has `_4` as type, which would be the `Typedef` with name `type`. What would be the xml output you would expect?

bradking replied:
Is it `ElaboratedType` because someone wrote `struct type` or `ns::type`?  Do we even need to inject this in the latter case?  IIRC the whole point of this was to tell apart `foo` and `struct foo` syntax.

iMichka commented:
You are right about the original purpose of this pull request. I added a check in Output.cxx, to filter out just the elaborated type specifiers we are interested in.

These 2 test have been reset to their original state (they are unmodified now): `Class-template-member-nontemplate-Typedef` and `qualified-type-name`.

I have also rebased on master. I do not remember if I pulled your tweaks, this was 7 days ago and I already checked out the code, but did not touch it since then. I just thought about it when force pushing. Please double check if I didn't remove your changes. If it is the case, I do apologise for this.

bradking commented:
Thanks.  It looks like you did preserve my updates.

bradking commented:
I updated your branch to revise the commit message.

iMichka commented:
Great :) Now I will have some work to do to add all this to pygccxml :)

bradking commented:
Nice work!
